pallets/mining/src/lib.rs

```diff
@@ -0,0 +1,376 @@
+        #[pallet::weight(100_000)]
+        pub fn add_minting_origin(origin: OriginFor<T>, who: T::AccountId) -> DispatchResultWithPostInfo {
+            T::AdminOrigin::ensure_origin(origin)?;
+            Self::do_add_minting_origin(who)?;
+            Ok(().into())
+        }
+
+        /// Remove Minting Origin to Mining Resource
+        #[pallet::weight(100_000)]
+        pub fn remove_minting_origin(origin: OriginFor<T>, who: T::AccountId) -> DispatchResultWithPostInfo {
+            T::AdminOrigin::ensure_origin(origin)?;
+            Self::do_remove_minting_origin(who)?;
+            Ok(().into())
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+}
+
+impl<T: Config> Module<T> {
+    pub fn bit_mining_resource_account_id() -> T::AccountId {
+        T::BitMiningTreasury::get().into_account()
+    }
+
+    fn bit_mining_resource_currency_id() -> SocialTokenCurrencyId {
+        T::BitMiningResourceId::get()
+    }
+
+    pub fn is_mining_origin(who: &T::AccountId) -> bool {
+        let minting_origin = Self::minting_origin(who);
+        minting_origin == Some(())
+    }
+
+    pub fn ensure_admin(o: T::Origin) -> DispatchResult {
+        T::AdminOrigin::try_origin(o)
+            .map(|_| ())
+            .or_else(ensure_root)?;
+        Ok(())
+    }
+
+    fn do_mint(
+        who: T::AccountId,
+        amount: Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        ensure!(
+            Self::is_mining_origin(&who),
+            Error::<T>::NoPermission
+        );
+
+        let mining_treasury = Self::bit_mining_resource_account_id();
+        //Deposit Bit mining to mining treasury
+        T::MiningCurrency::deposit(Self::bit_mining_resource_currency_id(), &mining_treasury, amount)?;
+
+        Self::deposit_event(Event::<T>::MiningResourceMinted(amount));
+
+        Ok(())
+    }
+
+    fn do_burn(
+        who: T::AccountId,
+        amount: Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+        ensure!(
+            Self::is_mining_origin(&who),
+            Error::<T>::NoPermission
+        );
+
+        let mining_treasury = Self::bit_mining_resource_account_id();
+        ensure!(
+            T::MiningCurrency::can_slash(Self::bit_mining_resource_currency_id(), &mining_treasury, amount),
+            Error::<T>::BalanceZero
+        );
+        //Deposit Bit mining to mining treasury
+        T::MiningCurrency::slash(Self::bit_mining_resource_currency_id(), &mining_treasury, amount);
+
+        Self::deposit_event(Event::<T>::MiningResourceBurned(amount));
+
+        Ok(())
+    }
+
+    fn do_deposit(
+        who: T::AccountId,
+        from: T::AccountId,
+        amount: Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        ensure!(
```

Comment:
Hi Justin, I'm not sure about the is_mining_origin part. Does this mean only allow mining_origin could deposit and withdraw?

Reply:
That's right. Only certain origin can trigger deposit and withdraw.